gh-86519: Add prefixmatch APIs to the re module (#31137)
python/cpython: Doc/library/re.rst

Reviewer: ```suggestion
search() vs. prefixmatch()
^^^^^^^^^^^^^^^^^^^^^^^^^^
```

   Compile a regular expression pattern into a :ref:`regular expression object
   <re-objects>`, which can be used for matching using its

>    :func:`~Pattern.prefixmatch` (:func:`~Pattern.match`),

Reviewer: ```suggestion
   :func:`~Pattern.prefixmatch`,
```

.. function:: match(pattern, string, flags=0)

> .. function:: prefixmatch(pattern, string, flags=0)

Reviewer: `prefixmatch` first = preferred:
```suggestion
.. function:: prefixmatch(pattern, string, flags=0)
.. function:: match(pattern, string, flags=0)
```

   If zero or more characters at the beginning of *string* match the regular
   expression *pattern*, return a corresponding :class:`~re.Match`.  Return
   ``None`` if the string does not match the pattern; note that this is
   different from a zero-length match.

   Note that even in :const:`MULTILINE` mode, this will only match at the

>    beginning of the string and not at the beginning of each line.

Reviewer: Make this distinction more obvious:

```suggestion
   .. note::

      Even in :const:`MULTILINE` mode, this will only match at the
      beginning of the string and not at the beginning of each line.
```

.. method:: Pattern.match(string[, pos[, endpos]])

> .. method:: Pattern.prefixmatch(string[, pos[, endpos]])

Reviewer: ```suggestion
.. method:: Pattern.prefixmatch(string[, pos[, endpos]])
.. method:: Pattern.match(string[, pos[, endpos]])
```

   If zero or more characters at the *beginning* of *string* match this regular
   expression, return a corresponding :class:`~re.Match`. Return ``None`` if the
   string does not match the pattern; note that this is different from a
   zero-length match.

   Note that even in :const:`MULTILINE` mode, this will only match at the
   beginning of the string and not at the beginning of each line.

   The optional *pos* and *endpos* parameters have the same meaning as for the
   :meth:`~Pattern.search` method. ::

      >>> pattern = re.compile("o")
      >>> pattern.prefixmatch("dog")     # No match as "o" is not at the start of "dog".
      >>> pattern.prefixmatch("dog", 1)  # Match as "o" is the 2nd character of "dog".
      <re.Match object; span=(1, 2), match='o'>
      >>> pattern.match("dog")           # Same as above.
      >>> pattern.match("dog", 1)        # Same as above.

>       <re.Match object; span=(1, 2), match='o'>

Reviewer: Encourage the longer name.
```suggestion
```

To see if a given string is a valid hand, one could do the following::

   >>> valid_hand_re = re.compile(r"^[a2-9tjqk]{5}$")
   >>> displaymatch(valid_hand_re.search("akt5q"))  # Valid.
   "<Match: 'akt5q', groups=()>"
   >>> displaymatch(valid_hand_re.search("akt5e"))  # Invalid.
   >>> displaymatch(valid_hand_re.search("akt"))    # Invalid.
   >>> displaymatch(valid_hand_re.search("727ak"))  # Valid.
   "<Match: '727ak', groups=()>"

That last hand, ``"727ak"``, contained a pair, or two of the same valued cards.
To match this with a regular expression, one could use backreferences as such::

   >>> pair_re = re.compile(r".*(.).*\1")
   >>> displaymatch(pair_re.match("717ak"))     # Pair of 7s.
   "<Match: '717', groups=('7',)>"
   >>> displaymatch(pair_re.match("718ak"))     # No pairs.

>    >>> displaymatch(pair_re.match("354aa"))     # Pair of aces.

Reviewer: I suggest only using `prefixmatch` in examples. We explain the double names in the main `Pattern.prefixmatch`/`Pattern.match` ref.
 
```suggestion
   >>> displaymatch(pair_re.prefixmatch("717ak"))     # Pair of 7s.
  "<Match: '717', groups=('7',)>"
   >>> displaymatch(pair_re.prefixmatch("718ak"))     # No pairs.
   >>> displaymatch(pair_re.prefixmatch("354aa"))     # Pair of aces.
```

   "<Match: '354aa', groups=('a',)>"

To find out what card the pair consists of, one could use the
:meth:`~Match.group` method of the match object in the following manner::

   >>> pair_re = re.compile(r".*(.).*\1")
   >>> pair_re.prefixmatch("717ak").group(1)
   '7'

   # Error because prefixmatch() returns None, which doesn't have a group() method:
   >>> pair_re.prefixmatch("718ak").group(1)
   Traceback (most recent call last):
     File "<pyshell#23>", line 1, in <module>
       pair_re.prefixmatch("718ak").group(1)
   AttributeError: 'NoneType' object has no attribute 'group'

   >>> pair_re.prefixmatch("354aa").group(1)
   'a'

> The examples above use :meth:`~Pattern.match` and :meth:`~Pattern.prefixmatch`

Reviewer 2: Given that we have the `_prefixmatch-vs-match:` section, it might be fine to just use `prefixmatch` and skip this paragraph

interchangeably because they are two names for the same method.
:meth:`~Pattern.prefixmatch` was added in Python 3.15 as a more explicit name;
use it when your code does not need to run on older Python versions.
:meth:`~Pattern.search` with a ``^`` or ``\A`` anchor is equivalent, but using
an explicit method name is clearer to readers of the code.
See :ref:`prefixmatch-vs-match` for more on this topic.

> 

Reviewer: ```suggestion
```

Python offers different primitive operations based on regular expressions:

+ :func:`re.prefixmatch`, also known under the less explicit name

>   :func:`re.match`, checks for a match only at the beginning of the string

Reviewer: ```suggestion
+ :func:`re.prefixmatch` checks for a match only at the beginning of the string
```

+ :func:`re.search` checks for a match anywhere in the string
  (this is what Perl does by default)
+ :func:`re.fullmatch` checks for entire string to be a match

For example::

>    >>> re.match("c", "abcdef")        # No match

Reviewer: ```suggestion
```

Regular expressions beginning with ``'^'`` can be used with :func:`search` to
restrict the match at the beginning of the string::

   >>> re.match("c", "abcdef")        # No match
   >>> re.prefixmatch("c", "abcdef")  # No match
   >>> re.search("^c", "abcdef")      # No match
   >>> re.search("^a", "abcdef")      # Match
   <re.Match object; span=(0, 1), match='a'>

Note however that in :const:`MULTILINE` mode :func:`match` only matches at the
beginning of the string, whereas using :func:`search` with a regular expression
beginning with ``'^'`` will match at the beginning of each line. ::

   >>> re.prefixmatch("X", "A\nB\nX", re.MULTILINE)  # No match
   >>> re.match("X", "A\nB\nX", re.MULTILINE)  # No match
   >>> re.search("^X", "A\nB\nX", re.MULTILINE)  # Match
   <re.Match object; span=(4, 5), match='X'>

.. _prefixmatch-vs-match:

prefixmatch() vs. match()
^^^^^^^^^^^^^^^^^^^^^^^^^

Why is the :func:`~re.match` function and method name being discouraged in

> favor of the longer :func:`~re.prefixmatch` spelling in very recent Python?

Reviewer: "very recent" will go out of date. 

```suggestion
favor of the longer :func:`~re.prefixmatch` spelling?
```

Many other languages have gained regex support libraries since regular
expressions were added to Python. However in the most popular of those, they
use the term *match* in their APIs to mean the unanchored behavior provided in
Python by :func:`~re.search`. Thus use of the plain term *match* can be
unclear to those used to other languages when reading or writing code and
not familiar with the Python API's divergence from what otherwise become the
industry norm.

Quoting from the Zen Of Python (``python3 -m this``): *"Explicit is better than
implicit"*. Anyone reading the name :func:`~re.prefixmatch` is likely to
understand the intended semantics. When reading :func:`~re.match` there remains
a seed of doubt about the intended behavior to anyone not already familiar with
this old Python gotcha.

We **do not** plan to deprecate and remove the older *match* name in this

> decade, if ever, as it has been used in code for over 25 years.

Reviewer: The "this decade" will eventually go out of date.
```suggestion
We **do not** plan to deprecate and remove the older *match* name,
if ever, as it has been used in code for over 25 years.
```

Regular expressions beginning with ``'^'`` can be used with :func:`search` to
restrict the match at the beginning of the string::

>    >>> re.match("c", "abcdef")        # No match

Reviewer: ```suggestion
```

   >>> re.prefixmatch("c", "abcdef")  # No match
   >>> re.search("^c", "abcdef")      # No match
   >>> re.search("^a", "abcdef")      # Match
   <re.Match object; span=(0, 1), match='a'>

Note however that in :const:`MULTILINE` mode :func:`match` only matches at the
beginning of the string, whereas using :func:`search` with a regular expression
beginning with ``'^'`` will match at the beginning of each line. ::

   >>> re.prefixmatch("X", "A\nB\nX", re.MULTILINE)  # No match
   >>> re.match("X", "A\nB\nX", re.MULTILINE)  # No match
   >>> re.search("^X", "A\nB\nX", re.MULTILINE)  # Match
   <re.Match object; span=(4, 5), match='X'>

.. _prefixmatch-vs-match:

prefixmatch() vs. match()
^^^^^^^^^^^^^^^^^^^^^^^^^

> Why is the :func:`~re.match` function and method name being discouraged in

Reviewer: ```suggestion
Why is the :func:`~re.match` function and method discouraged in
```
